lib/recursive-methods.rb

```diff
@@ -1,49 +1,117 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 0 || n == 1
+    return 1
+  elsif n < 0
+    raise ArgumentError
+  end
+  
+  until n == 1 do
+    return n * factorial(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(s)
```

Comment:
1.  Recursion means you have some space complexity by the system stack.
2.  You are creating a new string with each recursive call so this is actually O(n<sup>2</sup>)